File: crates/agrr-domain/src/cultivation_plan/dtos/crop_task_schedule_blueprint.rs
//! Blueprint copy DTOs.

use rust_decimal::Decimal;

#[derive(Debug, Clone, PartialEq)]
pub struct CropTaskScheduleBlueprintCopyInput {
    pub reference_crop_id_to_user_crop_id: Vec<(i64, i64)>,
}

impl CropTaskScheduleBlueprintCopyInput {
    pub fn from_map(map: impl IntoIterator<Item = (i64, i64)>) -> Self {
        Self {
            reference_crop_id_to_user_crop_id: map.into_iter().collect(),
        }
    }

    pub fn is_empty(&self) -> bool {
        self.reference_crop_id_to_user_crop_id.is_empty()
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct CropTaskScheduleBlueprintRow {
    pub agricultural_task_id: Option<i64>,
    pub source_agricultural_task_id: Option<i64>,
    pub stage_order: i32,
    pub stage_name: String,
    pub gdd_trigger: Option<Decimal>,
    pub gdd_tolerance: Option<Decimal>,
    pub task_type: String,
    pub source: String,
    pub priority: i32,
    pub amount: Option<Decimal>,
    pub amount_unit: Option<String>,
    pub description: Option<String>,
    pub weather_dependency: Option<String>,
    pub time_per_sqm: Option<Decimal>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct CropTaskScheduleBlueprintCreateAttrs {
    pub crop_id: i64,
    pub agricultural_task_id: Option<i64>,
    pub source_agricultural_task_id: Option<i64>,
    pub stage_order: i32,
    pub stage_name: String,
    pub gdd_trigger: Option<String>,
    pub gdd_tolerance: Option<String>,
    pub task_type: String,
    pub source: String,
    pub priority: i32,
    pub amount: Option<String>,
    pub amount_unit: Option<String>,
    pub description: Option<String>,
    pub weather_dependency: Option<String>,
    pub time_per_sqm: Option<String>,
}

File: frontend/src/app/domain/crops/crop-task-schedule-blueprint.ts
export interface CropTaskScheduleBlueprintRelatedTask {
  id: number;
  name: string;
  description?: string | null;
  is_reference?: boolean;
}

export interface CropTaskScheduleBlueprint {
  id: number;
  crop_id: number;
  agricultural_task_id: number | null;
  source_agricultural_task_id: number | null;
  stage_order: number;
  stage_name: string | null;
  gdd_trigger: number;
  gdd_tolerance: number | null;
  task_type: string;
  source: string;
  priority: number;
  amount: number | null;
  amount_unit: string | null;
  description: string | null;
  weather_dependency: string | null;
  time_per_sqm: number | null;
  name?: string | null;
  agricultural_task?: CropTaskScheduleBlueprintRelatedTask | null;
  created_at?: string | null;
  updated_at?: string | null;
}
